### ISSUE TITLE
Rescaling changes categorical column encodings

### ISSUE THREAD

**Issue by isilopekel** [automl/auto-sklearn#619]
Hello,   
On a very simple data set, I have observed that the categorical columns are also affected by the normalization. I am wondering if there is any reason behind the normalization of  the categorical columns. 

There are 2 problems:
1. Rescaling->Normalizer.py  applies **row-based** rescaling, which can have unintended effects on unrelated columns. See the example below. It changes the relation between the same features in different rows.

2. Even if other rescalers are used in a columnar way, it does not make any sense to change **0. 1. 0.** to  **-0.5 1. 0**.  In addition, it causes an overhead. 

To reproduce:   

Ptyhon version: 3.6.5, ubuntu 16.04
```python
import autosklearn.classification
import sklearn.model_selection
import sklearn.datasets
import sklearn.metrics
import pandas as pd
data = pd.DataFrame.from_dict({'col_1': [3,4,5,6]*4, 'col_2': [0,1,2,1]*4, "col_3": [20, 20, 10, 10]*4})  
label= pd.DataFrame.from_dict({"label": [9,9,8,8]*4})  

data_train, data_test, label_train, label_test = sklearn.model_selection.train_test_split(data, label, random_state=1)   
automl = autosklearn.classification.AutoSklearnClassifier(time_left_for_this_task=3600)  
automl.fit(data_train, label_train, feat_type=["Numerical", "Categorical", "Numerical"])  
test_predicted = automl.predict(data_test)  
``` 
To print the rescaling effect, adapt the abstract_scaling.py -> Transform function as follows:  

```python  
    def transform(self, X):  
        if self.preprocessor is None:  
            raise NotImplementedError()  
        print("Before scaling\n",  X)  
        transformed = self.preprocessor.transform(X)  
        print("Transformed:\n", transformed)  
        return transformed  
```  

Result: 
***
Normalizer Rescaling!  (this one has been chosen)
Before scaling
 [[ 1.  0.  0.  3. 20.]
 [ 1.  0.  0.  3. 20.]
 [ 0.  0.  1.  5. 10.]
 [ 0.  1.  0.  6. 10.]  --> 0 1 0 
 [ 0.  0.  1.  5. 10.]
 [ 1.  0.  0.  3. 20.]
 [ 0.  1.  0.  4. 20.] --> same 0 1 0
 [ 1.  0.  0.  3. 20.]] 
***
Transformed :
 [[0.04938648 0.         0.         0.14815944 0.9877296 ]
 [0.04938648 0.         0.         0.14815944 0.9877296 ]
 [0.         0.         0.08908708 0.4454354  0.89087081]
 [0.         0.08543577 0.         0.51261459 0.85435766]  ---> 1  is encoded with 0.085
 [0.         0.         0.08908708 0.4454354  0.89087081]
 [0.04938648 0.         0.         0.14815944 0.9877296 ]
 [0.         0.04897021 0.         0.19588084 0.97940421]   ---> 1 is encoded with 0.048
 [0.04938648 0.         0.         0.14815944 0.9877296 ]] 

Also all 1's get a different float assigned in the first 3 columns.

Thank you! Isil

**Comment by mfeurer** [comment]
Hi @isilopekel I fully agree that this is suboptimal and I think no one though about that it might actually be harmful when normalizing data. Let me give you a few answers:

1. It's currently done this way because we do not yet implement the feature transformation inside Auto-sklearn because our code to generate the configuration space only supports linear pipelines. This will require quite some effort and is not on our TODO list. Any help here would be appreciated, but that'll require quite some effort.
2. Is there any paper/research/resource which shows that normalizing one hot encoded data is bad? While it initially sounds like a bad idea, I wonder whether it has any effect in practice for the datasets where normalization would be applied.

**Comment by isilopekel** [comment]
Hi Matthias,

Thank you for your answers. 
1) Yes, we have tried to fix it but we could not keep the original sklearn logic fully. Some preprocessing steps have been disabled.  Otherwise, we would be glad to contribute to the repo. A linear pipeline is very useful but the OHE in the first step might change the number of columns in the input and the following steps cannot capture it anymore. All steps are applied blindly on to the whole data. Difficult to change.

2) I am not aware of any papers. However, I try to be careful if I need to deal with word-embeddings. The dimensions are somehow fragile, and ML encoders are not always conditioned to to map the input space into a clearly separable output space. I can imagine that the differences in the embedding vectors can be very very fine.

3) Before we close this issue, i would like to ask if it makes sense to include a row-based normalizer in the rescaler set? If yes, could you please let me understand it? 

Thank you ! 
Best 
Isil

**Comment by mfeurer** [comment]
Regarding 3: we have found this to be very helpful for word count and TF/IDF datasets where it improved over not scaling the data.

**Comment by mfeurer** [comment]
I'm not sure what to do about the others. I'm afraid we need to wait until we find someone to spent some time on this.

**Comment by isilopekel** [comment]
Interesting. Tf-Idf is already a normalized representation. I would not expect a contribution from normalizing. 
If the dataset is only composed of bag-of-words and no other features,  this can make a contribution indeed. 

Thank you very much. I am closing the issue now.

**Comment by mfeurer** [comment]
Then it probably only was bag of words. It's already a few years ago...
